### ISSUE TITLE
man page generation can result in long file names, even using SHORT_NAMES (Origin: bugzilla #703547)

### ISSUE THREAD

**Issue by doxygen** [doxygen/doxygen#5229]
status NEEDINFO severity *major* in component *general* for *---*
Reported in version *1.8.4* on platform *Other*
Assigned to: Dimitri van Heesch

On 2013-07-03 14:25:39 +0000, chris.beer@detica.com wrote:

> When executing Doxygen in deeply nested paths, file names too long for the OS to handle can result. This does not happen in version 1.6.1.

On 2013-07-05 20:31:26 +0000, Dimitri van Heesch wrote:

> Can you please add a self contained example that produces such long path, and indicate which OS & filesystem you are using.

On 2013-07-08 16:54:13 +0000, chris.beer@detica.com wrote:

> OS used is CentOS 6.3. This bug is specific to manpage generation.
> 
> mangen.cpp builds a unique filename for each level of directory using the full path to that directory, substituting special characters. For example /a/short/path/ becomes a_short_path_.3. Try a path more than 500 characters long, for example /a/very/long/path/a/very/long/path/a/very/long/path/.../a/very/long/path. The resulting filename is not capped in length and therefore can exceed OS limitations (in this case 500 chars).
> 
> These files with names generated based on full path did not exist back in 1.6.1, hence there was no issue.

**Comment by sawatts** [comment]
### Example

Error which ended our build when generating docs:
```
error: Could not open file <builddir>/man/man3/BloombergLP_pasconv_Any_Converter_ bsl_string, Out, typename bsl_enable_if_ pasmf_EnableIfHelpers_And_ pasmf_EnableIfHelpers_Or3_ bsl_is_same_ Out, double __value, bsl_is_same_ Out, float __value, bsl_is_integral_ Out __value __value, pasmf_EnableIfHelpers_Not_ IsCharVariant_ Out __value __value __value __type _.3libpas for writing
```

Try to list (`ls`) the given filename (in case it already existed) gave the following:
```
$ ls "man/man3/BloombergLP_pasconv_Any_Converter_ bsl_string, Out, typename bsl_enable_if_ pasmf_EnableIfHelpers_And_ pasmf_EnableIfHelpers_Or3_ bsl_is_same_ Out, double __value, bsl_is_same_ Out, float __value, bsl_is_integral_ Out __value __value, pasmf_EnableIfHelpers_Not_ IsCharVariant_ Out __value __value __value __type _.3libpas"
ls: cannot access 'man/man3/BloombergLP_pasconv_Any_Converter_ bsl_string, Out, typename bsl_enable_if_ pasmf_EnableIfHelpers_And_ pasmf_EnableIfHelpers_Or3_ bsl_is_same_ Out, double __value, bsl_is_same_ Out, float __value, bsl_is_integral_ Out __value __value, pasmf_EnableIfHelpers_Not_ IsCharVariant_ Out __value __value __value __type _.3libpas': File name too long
```

Note error message "File name too long".

### Suggestion

Could parameterized types and overloaded methods be accessed via their base names, without parameters, and some selection included after that?

### What Where We Doing?

We use a custom **cmake** module to simplify the large number of projects we have.  This includes building the documentation set using **doxygen** and publishing it to an internal area.  We include **man** page output as an option, which can then be added to the development package for a library.  We add the library name to the man group to avoid conflicts.

We would also like to produce man pages for other groups - such as command line tools (1), and file formats (5) - which are part of the docset but appear in man3 (if at all).

**Comment by albert-github** [comment]
@sawatts 
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.  
- Please also specify the full doxygen version used (`doxygen -v`).

**Comment by sawatts** [comment]
### Item being Documented

I believe that the following is the item being documented at this point. It is a class template specialisation:
```c++
template <typename Out>
struct Converter
<
	std::string,
	Out,
	typename std::enable_if
	<
		xxxmf::EnableIfHelpers::And
		<
			xxxmf::EnableIfHelpers::Or3
			<
				std::is_same<Out,double>::value,
				std::is_same<Out,float>::value,
				std::is_integral<Out>::value
			>::value,
			xxxmf::EnableIfHelpers::Not
			<
				IsCharVariant<Out>::value
			>::value
		>::value
	>::type
>
```
Which is a specialisation of:
```c++
template 
<
	typename Inp,
	typename Out,
	typename Enable = void
>
struct Converter;
```

I will try to find time to make a demo based on this example - but can't do it just now.  Will add a link to this ticket against our internal ticket.

**Comment by albert-github** [comment]
Doxyfile would probably easy to provide (can be one of a full project).

**Comment by sawatts** [comment]
### Original Project Doxyfile

[Doxyfile.zip](https://github.com/doxygen/doxygen/files/5874695/Doxyfile.zip)

This was generated by **cmake** 3.19.3 using the `FindDoxygen` module. Additional configuration variables controlled by our module.

**Comment by albert-github** [comment]
@sawatts Thanks for the file  didn't bring what I was hoping, so indeed a , small, self contained example is requested.

**Comment by sawatts** [comment]
### Self Contained Example

[doxygen-5229.zip](https://github.com/doxygen/doxygen/files/5875008/doxygen-5229.zip)

Consists of one (1) C++ header file containing the definitions corresponding to the suspect code given above.

One `Doxyfile` created from the original project file.

One `DOXYGEN.LOG` as the console output (std+err) from running `doxygen` (1.8.15) from the same dir as the `Doxyfile`, and showing the same error as given above.

Output is to an `out` subdirectory, containing `html` and `man`.

uname -a:
```
Linux riskldevnj2 2.6.32-642.6.2.el6.x86_64 #1 SMP Mon Oct 24 10:22:33 EDT 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by albert-github** [comment]
With the given example I was able to reproduce the problem with the current master version (under windows).
Notes:
- for the example it is **not necessary** to have `CREATE_SUBDIRS=YES`
- to get the html output we have to disable the `GENERATE_MAN=YES` because as soon as the error occurs doxygen is stopped in a regular way (but this is hidden due to the `QUIET=YES`)
- In the HTML output we see on the page: `out/html/annotated.html`
  the line:
  ```
  Converter<  std::string, OUTPUT, typename std::enable_if<  helpers::enableifhelpers::And< helpers::enableifhelpers::Or3<  std::is_same< OUTPUT, double >::value, std::is_same< OUTPUT,  float >::value, std::is_integral< OUTPUT >::value >::value,  helpers::enableifhelpers::Not< IsCharVariant< OUTPUT >::value  >::value >::value >::type >
  ```
  that refers to
  ```
  out/html/structdoxygen5229_1_1sawatts_1_1example_1_1Converter_3_01std_1_1string_00_01OUTPUT_00_01typenamef892964bb28615e2d2ecbb8f3122cda3.html
  ```
  whilst the man page wants to create the file:
  ```
  out/man/man3/doxygen5229_sawatts_example_Converter_ std_string, OUTPUT, typename std_enable_if_ helpers_enableifhelpers_And_ helpers_enableifhelpers_Or3_ std_is_same_ OUTPUT, double __value, std_is_same_ OUTPUT, float __value, std_is_integral_ OUTPUT __value __value, helpers_enableifhelpers_Not_ IsCharVariant_ OUTPUT __value __value __value __type _.3example

**Comment by albert-github** [comment]
I can see that this is a problem, though to be able to find a solution I have a few questions regarding addressing / working with the man pages (I only used it for normal functions like `printf`, `strcat` etc.). I manipulated doxygen a bit so it "skipped" the offending file and didn't crash (but this is not a real solution as it is a dirty hack in a very general routine and can have negative effects at other places).
- When one needs the information about e.g. print one gives `man 3 printf`
  -  how would you address `IsCharVariant` i.e. `doxygen5229::sawatts::example::IsCharVariant`, I think the file name here is `doxygen5229_sawatts_example_IsCharVariant.3` (I removed everywhere, for convenience, the `example`).
  - how would you address `IsCharVariant<char>`, I think the file name here is `doxygen5229_sawatts_example_IsCharVariant_ char _.3`
- In the file `man converter.h` there are a lot of functions / classes / Namespaces how to jump to the detailed description of such an item (as test I added some "brief" and "detailed" information)

**Comment by sawatts** [comment]
In our case the **Doxygen** is generated by a framework which sits over hundreds of projects, mostly poorly documented.  The main goal is the _HTML_ output, but it would be good to add _man_ pages to their dev packages.


The **man** system has problems with the complex types we now get from languages such as _C++_, which where not around when it was designed.

There are examples which use `<namespace>::<symbol>` man page filenames - for example, on _RHEL_ and _Ubuntu_.  Advantage to keeping the same notation as used in the language (if possible) otherwise users have to guess the transform.

_IMHO_ I would group overloads under the same manpage - both functions and template types with deductions.

So don't encode function parameters into the page name, but group all overloads of a function within the same page - so page _foo::bar_ documents `foo::bar(int)`, `foo::bar(float)` _etc_.

The same should then be true for template parameters and deductions.  So in the sample case there would be a manpage for `doxygen5229::sawatts::example::Converter` which includes the documentation for the generic `Converter` template class, and the specialisation with the deep `enable_if` deduction. 

#### TL;DR: 

- `InCharVariant` should give manpage `doxygen5229::sawatts::example::IsCharVariant.3`.
- `InCharVariant<char>` specialisation should be documented within the same man page as `InCharVariant`.
- `man converter.h` - I didn't know this was generated, but should be in `man5` rather than `man3`.  Since there isn't a "hyperlink" mechanism between man pages, the usual reference is to list in the `SEE ALSO` section.

Note that `apropos` is your friend when looking for the right man page; so `apropos Coverter` would list `doxygen5229::sawatts::example::Converter (3example)` along with any other pages which match the term.

**Comment by Maxime2** [comment]
Perhaps a temporary solution would be in introducing MAN_EXCLUDE_SYMBOLS configuration option similar to EXCLUDE_SYMBOLS but affecting only man page generation as HTML generated documentation has no such a problem when SHORT_NAMES option is enabled. So the affected symbols would be described in HTML docs while be missed in man pages.
